### PR TITLE
Refactor Relayer Game

### PR DESCRIPTION
**Pull request by AurevoirXavier**
- [x] new api
- [x] new storage
- [x] new settle algorithm
- [x] new verification
- [ ] code docs (70%)
- [x] migrations
- [x] types json
- [ ] relayer game unit tests (50%)
- [ ] ethereum relay unit tests (test data needed @clearloop)
- [ ] ethereum backing unit tests (test data needed @clearloop)

### CODE REVIEW

**Comment by HackFisher** [comment]
It's more clear now, I like these changes.

Please make the api code changes today, as dj need time for testing before mainet launch.

**Comment by wuminzhe** [comment]
`fn extend proposal` may need a MMRProof:
```
enum Relay Stuff<Header, MMRRoot, MMRProof> {
	Header(Header),
	MMRRoot(MMRRoot),
        MMRProof(MMRProof),
	Other(Vec<u8>),
}
```

File: primitives/relay/src/lib.rs [hunk 1]
```diff
@@ -8,238 +8,133 @@ use core::fmt::Debug;
...
 
-	fn approve_pending_header(
-		pending: <Self::HeaderThing as HeaderThing>::Number,
+	/// Verify a specify proposal
+	///
+	/// Proofs is a `Vec` because the sampling function might give more than 1 sample points,
+	/// so need to verify each sample point with its proofs
+	fn complete_proofs(
+		proposal_id: ProposalId<Self::GameId>,
+		proofs: Vec<Self::Proofs>,
```
Comment (wuminzhe) [inline primitives/relay/src/lib.rs:85]: proofs_list maybe better

**Review by wuminzhe — APPROVED** [review]

File: frame/bridge/relayer-game/src/lib.rs [hunk 2]
```diff
@@ -88,93 +94,85 @@ impl WeightInfo for () {}
...
-			get(fn samples_of_game)
-			: map hasher(blake2_128_concat) TcBlockNumber<T, I>
-			=> Vec<Vec<TcBlockNumber<T, I>>>;
-
-		/// The closed rounds which had passed the challenge time at this moment
+			get(fn last_confirmed_block_id_of)
+			: map hasher(identity) GameId<T, I>
+			=> RelayBlockId<T, I>;
+
+		pub RoundCounts
```
Comment (HackFisher) [inline frame/bridge/relayer-game/src/lib.rs:150]: What is RoundCounts?

File: frame/bridge/relayer-game/src/lib.rs [hunk 3]
```diff
@@ -88,93 +94,85 @@ impl WeightInfo for () {}
...
+			: map hasher(identity) GameId<T, I>
+			=> u32;
+
+		/// All the closed games here
+		///
+		/// Closed games won't accept any proposal
+		pub ClosedGames
+			get(fn game_closed_at_of)
+			: map hasher(identity) GameId<T, I>
+			=> BlockNumber<T>;
```
Comment (HackFisher) [inline frame/bridge/relayer-game/src/lib.rs:161]: What's the meaning of value, the type of which is ```BlockNumber<T>```

File: frame/bridge/relayer-game/src/lib.rs [hunk 4]
```diff
@@ -88,93 +94,85 @@ impl WeightInfo for () {}
...
+
+		/// All the closed rounds here
+		///
+		/// Record the closed rounds endpoint which use for settlling or updating
 		pub ClosedRounds
 			get(fn closed_rounds_at)
-			: map hasher(blake2_128_concat) BlockNumber<T>
-			=> Vec<(GameId<TcBlockNumber<T, I>>, Round)>;
+			: map hasher(identity) BlockNumber<T>
```
Comment (HackFisher) [inline frame/bridge/relayer-game/src/lib.rs:168]: Please give more docs about the storage key and value.

Comment (HackFisher) [reply frame/bridge/relayer-game/src/lib.rs:168]: The key is the time(block number) to close the rounds?

File: frame/bridge/relayer-game/src/lib.rs [hunk 5]
```diff
@@ -187,66 +185,168 @@ decl_module! {
...
-		// 	https://github.com/darwinia-network/darwinia-common/issues/253
-		fn on_finalize(block_number: BlockNumber<T>) {
-			let closed_rounds = <ClosedRounds<T, I>>::take(block_number);
-
-			// `closed_rounds` MUST NOT be empty after this check; qed
-			if closed_rounds.len() != 0 {
-				// TODO: handle error
-				let _ = Self::settle(block_number, closed_rounds);
+		#[weight = 100_000_000]
+		pub fn update_games_unsigned(origin, game_ids: Vec<GameId<T, I>>) {
```
Comment (HackFisher) [inline frame/bridge/relayer-game/src/lib.rs:201]:
```suggestion
		pub fn settle_games_unsigned(origin, game_ids: Vec<GameId<T, I>>) {
```

File: frame/bridge/relayer-game/src/lib.rs [hunk 6]
```diff
@@ -187,66 +185,168 @@ decl_module! {
...
+
+	// 	Ok(proposal)
+	// }
+
+	pub fn update_timer_of_game_at(game_id: &GameId<T, I>, round: u32) {
+		let propose_time = T::RelayerGameAdjustor::propose_time(round);
+		let complete_proofs_time = T::RelayerGameAdjustor::complete_proofs_time(round);
+
+		<ClosedGames<T, I>>::insert(game_id, propose_time);
+		<ClosedRounds<T, I>>::append(propose_time + complete_proofs_time, game_id);
```
Comment (HackFisher) [inline frame/bridge/relayer-game/src/lib.rs:332]:
```suggestion
		<GameProposeEndTime<T, I>>::insert(game_id, propose_time);
		<GamesToSettle<T, I>>::append(propose_time + complete_proofs_time, game_id);
```

File: frame/bridge/relayer-game/src/lib.rs [hunk 7]
```diff
@@ -187,66 +185,168 @@ decl_module! {
...
+	}
+
+	pub fn update_games_at(game_ids: Vec<GameId<T, I>>, moment: BlockNumber<T>) -> DispatchResult {
```
Comment (HackFisher) [inline frame/bridge/relayer-game/src/lib.rs:335]:
```suggestion
	pub fn settle_games_at(game_ids: Vec<GameId<T, I>>, moment: BlockNumber<T>) -> DispatchResult {
```

File: frame/bridge/relayer-game/src/lib.rs [hunk 8]
```diff
@@ -264,581 +364,352 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
...
+				|RelayProposal { relayer, bond, .. }, _, _| {
+					Self::slash_on(relayer, *bond);
+				},
+			);
 		}
+	}
 
-		for (evil, (evil_bonds, honesties_map)) in evils_map {
-			Self::update_bonds_with(&evil, |old_bonds| old_bonds.saturating_sub(evil_bonds));
+	pub fn settle_with_challenge(game_id: &GameId<T, I>, winning_proposal: RelayProposalT<T, I>) {
```
Comment (HackFisher) [inline frame/bridge/relayer-game/src/lib.rs:477]: // Settle a game after dispute resolution result outcome with one winning proposal

File: frame/bridge/relayer-game/src/lib.rs [hunk 9]
```diff
@@ -264,581 +364,352 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
...
-				samples.concat(),
-				samples[samples.len() - 1].clone(),
-			));
-		});
-	}
+			proposal
+		};
+		let round = 0;
+
+		<ActiveGames<I>>::mutate(|count| *count += 1);
```
Comment (HackFisher) [inline frame/bridge/relayer-game/src/lib.rs:624]: Why auto increase the counter if it is joining an existing game?

**Review by HackFisher — CHANGES_REQUESTED** [review]

File: frame/bridge/relayer-game/src/lib.rs (hunk 2 above)
Comment (AurevoirXavier) [reply frame/bridge/relayer-game/src/lib.rs:150]: Total rounds count of a game

File: frame/bridge/relayer-game/src/lib.rs (hunk 3 above)
Comment (AurevoirXavier) [reply frame/bridge/relayer-game/src/lib.rs:161]: Game close at this moment, won't accept any propsal

File: frame/bridge/relayer-game/src/lib.rs (hunk 4 above)
Comment (AurevoirXavier) [reply frame/bridge/relayer-game/src/lib.rs:168]: Settle or update will be scheduled which will start at this moment

File: frame/bridge/relayer-game/src/lib.rs (hunk 5 above)
Comment (AurevoirXavier) [reply frame/bridge/relayer-game/src/lib.rs:201]: Not just settle here

Would trigger a new round to update some games

File: frame/bridge/relayer-game/src/lib.rs (hunk 9 above)
Comment (AurevoirXavier) [reply frame/bridge/relayer-game/src/lib.rs:744]: Can't get the active games count without this, unless iterate the propsals but that's too expensive

Comment (AurevoirXavier) [reply frame/bridge/relayer-game/src/lib.rs:744]: And `propose` only accept new block

If join an existing game, must call `extend_proposal`

File: primitives/relay/src/lib.rs [hunk 10]
```diff
@@ -8,238 +8,127 @@ use core::fmt::Debug;
...
-		pending: <Self::HeaderThing as HeaderThing>::Number,
+	/// Once there're different opinions in a game,
+	/// chain will ask relayer to submit more samples
+	/// to help the chain make a on chain arbitrate finally
+	fn extend_proposal(
+		relayer: Self::Relayer,
+		samples: Vec<Self::RelayParcel>,
+		extended_relay_proposal_id: RelayProposalId<Self::GameId>,
+		proofses: Option<Vec<Self::Proofs>>,
 	) -> DispatchResult;
```
Comment (clearloop) [inline primitives/relay/src/lib.rs:105]: Still considering using `RelayParcelWithProof` here

```rust
struct RelayParcelWithProof {
  parcel: RelayParcel,
  proofs: Option<Proofs>
}

fn extend_proposal(
  relayer: Self::Relayer,
  samples: Vec<Self::RelayParcelWithProof>,
  extended_relay_proposal_id: RelayProposalId<Self::Proofs>
)
```

Fewer arguments might be better for the design of the API.

Comment (wuminzhe) [reply primitives/relay/src/lib.rs:105]: I prefer without a struct, but the 'proofses' look weird, maybe 'proofs_list' or just 'proofs', a proof is a proof of a sample.

Comment (AurevoirXavier) [reply primitives/relay/src/lib.rs:105]: That's what exactly the API's job.

If there's `option_proofs` then I just need one `if let Some()`.

But if there's a `Vec<RelayParcelWithProof>` then I've to `for relay_pracel_with_proof in samples { if let Some() }` for every sigle option filed.

**Comment by HackFisher** [comment]
@HackFisher What's the todo list for this PR? Can we review and merge now?

**Comment by AurevoirXavier** [comment]
> AurevoirXavier What's the todo list for this PR? Can we review and merge now?

See the description.

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 11]
```diff
@@ -129,21 +137,35 @@ darwinia_support::impl_genesis! {
 }
 decl_storage! {
 	trait Store for Module<T: Trait> as DarwiniaEthereumRelay {
-		/// Confirmed Ethereum Headers
-		pub ConfirmedHeaders
-			get(fn confirmed_header)
-			: map hasher(identity) EthereumBlockNumber => Option<EthereumHeaderThing>;
+		/// Relaied ethereum parcel
+		pub RelaiedParcels
```
Comment (HackFisher) [inline frame/bridge/ethereum/relay/src/lib.rs:141]:
```suggestion
		pub RelayedParcels
```

Spelling mistakes?

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 12]
```diff
@@ -156,19 +178,19 @@ decl_storage! {
...
-			ConfirmedBlockNumbers::mutate(|numbers| {
+			BestRelaiedBlockNumber::put(genesis_header.number);
+			RelaiedBlockNumbers::mutate(|numbers| {
 				numbers.push(genesis_header.number);
 
-				ConfirmedHeaders::insert(
+				RelaiedParcels::insert(
 					genesis_header.number,
-					EthereumHeaderThing {
+					EthereumRelayParcel {
```
Comment (HackFisher) [inline frame/bridge/ethereum/relay/src/lib.rs:187]:
```suggestion
					EthereumHeaderParcel {
```

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 13]
```diff
@@ -191,41 +213,97 @@ decl_module! {
...
+			ensure_signed(origin)?;
+
+			T::RelayerGame::complete_relay_proofs(proposal_id, ethereum_relay_proofs)?;
+		}
+
 		#[weight = 0]
-		pub fn submit_proposal(origin, proposal: Vec<EthereumHeaderThingWithProof>) {
+		fn extend_proposal(
+			origin,
+			game_sample_points: Vec<EthereumRelayParcel>,
```
Comment (HackFisher) [inline frame/bridge/ethereum/relay/src/lib.rs:265]: Multiple game_sample_points make things more complex.

@clearloop @wuminzhe Need further review

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 14]
```diff
@@ -191,41 +213,97 @@ decl_module! {
...
+				relayer,
+				game_sample_points,
+				extended_ethereum_relay_proposal_id,
+				optional_ethereum_relay_proofs
+			)?;
 		}
 
 		#[weight = 100_000_000]
-		pub fn approve_pending_header(origin, pending: EthereumBlockNumber) {
+		pub fn approve_pending_relay_parcel(origin, pending_relay_block_id: EthereumBlockNumber) {
```
Comment (HackFisher) [inline frame/bridge/ethereum/relay/src/lib.rs:280]:
```suggestion
		pub fn approve_pending_header_parcel(origin, pending_relay_block_id: EthereumBlockNumber) {
```

File: frame/bridge/ethereum/relay/src/lib.rs [hunk 15]
```diff
@@ -266,41 +346,57 @@ decl_module! {
 			<ReceiptVerifyFee<T>>::put(new);
 		}
 
-		/// Remove the specific malicous block
+		/// Remove the specific malicous relay parcel
 		#[weight = 100_000_000]
-		pub fn remove_confirmed_block(origin, number: EthereumBlockNumber) {
+		pub fn remove_relaied_parcel_of(origin, relaied_block_number: EthereumBlockNumber) {
```
Comment (HackFisher) [inline frame/bridge/ethereum/relay/src/lib.rs:351]:
```suggestion
		pub fn remove_confirmed_parcel_of(origin, relaied_block_number: EthereumBlockNumber) {
```

```Confirmed``` is more straight forward word

File: frame/bridge/ethereum/relay/src/mock.rs [hunk 16]
```diff
@@ -169,13 +169,13 @@ impl RelayerGameProtocol for UnusedRelayerGame {
 }
 
 pub fn proposal_of_game_with_id(
-	game_id: u64,
+	relay_block_id: u64,
```
Comment (HackFisher) [inline frame/bridge/ethereum/relay/src/mock.rs:172]: relay and block is a word neutral and ambiguous, not a good term IMO.

File: frame/bridge/relayer-game/src/lib.rs [hunk 17]
```diff
@@ -88,93 +90,110 @@ impl WeightInfo for () {}
...
+		/// Pending Relay Parcel - NOT EXISTED
+		PendingRelayParcelNE,
 	}
 }
 
 decl_storage! {
 	trait Store for Module<T: Trait<I>, I: Instance = DefaultInstance> as DarwiniaRelayerGame {
-		/// All the proposals here per game
+		/// Active games' relay block ids
+		pub BlocksToRelay get(fn blocks_to_relay): Vec<RelayBlockId<T, I>>;
```
Comment (HackFisher) [inline frame/bridge/relayer-game/src/lib.rs:139]:
```suggestion
		pub BlocksToResolve get(fn blocks_to_relay): Vec<RelayBlockId<T, I>>;
```

Game (断言, Relay?)
Verification Game(Dispute Resolution)

**Review by HackFisher — CHANGES_REQUESTED** [review]

**Review by HackFisher — APPROVED** [review]
LGTM for the new API, for tests and mocks, considering the constraint of time, let's leave it in next PR.

**Pull request merged**